test/test_dbus.c

```diff
@@ -153,6 +154,11 @@ int main(int argc, char *argv[])
                       dbus_message_iter_recurse(&dict, &val);
                       dbus_message_iter_get_basic(&val, &str);
                       printf("Value = %s\n", str);
+                    } else if (strcmp(property, ZBAR_SIGNAL_BINARY_DATA) == 0) {
```

Comment:
It sounds hacky to test for a property here. I mean, the D-Bus listener should not rely on anything except on the received messages.  See, the dbus testing code: it is possible to test if the received message is a string or not:

    if (DBUS_TYPE_STRING != dbus_message_iter_get_arg_type(&dict))

The right way seems to have an else clause there for the bytearray type.

Reply:
This line is checking the type of the key. Strings are the only valid keys. The `org.linuxtv.Zbar1.Code` signal is defined as a dictionary with string keys and variant values:

```C
dbus_message_iter_open_container(&args, DBUS_TYPE_ARRAY, "{sv}", &dict);
```

The types of the values are currently not checked. I wrote my code following the pattern of the surrounding code. So currently the code checks the key and recurses into the variant value without checking its type. Before this patch, only string values were possible. I agree that it'd be wise to add some type checks.

Since the nature of the data can be distinguished by checking its type, do we need a new `BinaryData` key? It should be possible to have `Data` contain both strings and byte arrays and clients should be able to handle both types.